firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/messages.txt

```diff
@@ -0,0 +1,207 @@
+/* Write function
+Scenario: Message: The write method correctly generates an encoded body from the protobuf message descriptor payload
+
+  GIVEN("A Message object default-constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes as its class template parameters")
+  // assertions carried out in given:
+  // message class type member is 0
+
+    // setup:
+    // message class type member set to 5
+    // payload of parameters_request is set
+    // message descriptor array containing one element corresponding to Unknown message
+
+    WHEN("Write method is called on a message object with it's value of the payload tag greater than the size of the descriptor array")
+      THEN("The write method reports invalid type status")
+      THEN("After the write method is called, the type member remains unchanged")
+      THEN("The payload.values field data remains unchanged")
+      THEN("The output buffer remains empty")
+
+    // setup:
+    // message class type member is unchanged
+    // payload tag is set to Unknown
+
+    WHEN("Write method is called on a message object where it's payload tag value is such that it corresponds to Unrecognized message descriptor in descriptor array")
+      THEN("The write method reports invalid type status")
+      THEN("After the write method is called, the type member remains unchanged")
+      THEN("The payload.values field data remains unchanged")
+      THEN("The output buffer remains empty")
+
+  GIVEN("A Message object default-constructed with StateSegment Taggedunion, default message types and a max size of 126 bytes as its class template parameters")
+    // setup:
+    // sensor_measurements.flow = 30;
+    // sensor_measurements.fio2 = 85;
+    // sensor_measurements.spo2 = 72;  
+    WHEN("The payload is a sensor measurements message and write method is called")
+      THEN("The write method reports ok status")
+      THEN("After the write method is called, the type member is set to the value of the message payload tag")
+      THEN("The type field of the body's header correctly stores the value of message type obtained from the message payload tag")
+      THEN("The output buffer has an expected sequence of 16 bytes")
+
+  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 508 bytes")
+    // setup:
+    // parameters.fio2 = 60;
+    // parameters.mode = VentilationMode_hfnc;
+    // parameters.ventilating = true;
+    WHEN("The parameters message data is written")
+      THEN("The write method reports ok status")
+      THEN("After the write method is called, the type member is set to the value of the message payload tag")
+      THEN("The type field of the body's header correctly stores the value of message type obtained from the message payload tag")
+      THEN("The output buffer has an expected sequence of 10 bytes")
+
```

Comment:
We're still missing a test for the `write` method returning `invalid_length`, which is one of the three possible error codes returned by this method. I believe there are two cases for this: 1) less than header size, and 2) greater than header size but less than the required size (e.g. for the sensor_measurements test which should generate a 16-byte output, setting max_size to 14 bytes).

If we're unable to generate tests which trigger `invalid_encoding`, we should add a comment to that effect in the test files, and explaining what we tried and why we think we won't be able to generate such a test.

Reply:
> We're still missing a test for the `write` method returning `invalid_length`, which is one of the three possible error codes returned by this method. I believe there are two cases for this: 1) less than header size, and 2) greater than header size but less than the required size (e.g. for the sensor_measurements test which should generate a 16-byte output, setting max_size to 14 bytes).

added.

> If we're unable to generate tests which trigger `invalid_encoding`, we should add a comment to that effect in the test files, and explaining what we tried and why we think we won't be able to generate such a test.

will do.